Add xmlLoadString function

Lpsd opened the pull request:
Currently, if you want to read XML data that is compatible with the other xml* functions, you need to create a file with the XML data (using `fileCreate`), then read that file using `xmlLoadFile`

This will allow you to read XML data via string using `xmlLoadString(xmlString)`

Example (client):

```lua
function init()
	print("Loading XML String [client]")

	local rootNode = xmlLoadString("<animals test='x'><monkey name='crosroad'></monkey> <turtle name='luxy'></turtle></animals>")

	if rootNode then
		local rootAttributes = xmlNodeGetAttributes(rootNode)
		print("Root Node", "Name: "..xmlNodeGetName(rootNode),  "Attributes :"..toJSON(rootAttributes))
		
		local children = xmlNodeGetChildren(rootNode)
		
		for i, childNode in ipairs(children) do
			local attributes = xmlNodeGetAttributes(childNode)
			
			print("Child #"..i, "Name: "..xmlNodeGetName(childNode), "Attributes :"..toJSON(attributes))
		end
	end
end
addEventHandler("onClientResourceStart", resourceRoot, init)
```

My only concerns in this commit are the following (someone who actually knows what they are doing should verify this):

Shared/XML/CXMLNodeImpl.cpp:17
Shared/XML/CXMLNodeImpl.h:62

First time looking into MTA's code, and hardly worked with C++ too, so let me know of anything dumb in my commit!

Note: I have been made aware of a branch which is ready to test, pugixml, so more than anything I am hoping this PR will speed that process up ;)

Lpsd committed:
xmlLoadString initial

Lpsd committed:
Fix returning xml-node

Lpsd committed:
Update server CLuaMain::ParseString

qaisjp commented on Server/mods/deathmatch/logic/lua/CLuaMain.h:
```diff
@@ -67,6 +67,7 @@ class CLuaMain            //: public CClient
     CMapManager*     GetMapManager() const { return m_pMapManager; };
 
     CXMLFile*     CreateXML(const char* szFilename, bool bUseIDs = true, bool bReadOnly = false);
+	CXMLNode*     ParseString(const char* strXmlContent);
```
```suggestion
    CXMLNode*     ParseString(const char* strXmlContent);
```

qaisjp commented:
What happens if someone calls xmlSaveFile on this node?

qaisjp committed:
Fix spacing issue

Co-Authored-By: Lpsd <40902730+Lpsd@users.noreply.github.com>

Lpsd commented:
I'm going to look at this shortly - I'll ensure that file-related XML functions do not work with xml-nodes created via xmlLoadString

Lpsd commented:
This should now be fixed.

```lua
function init()
	print("Loading XML String [client]")

	local rootNode = xmlLoadString("<animals test='x'><monkey name='crosroad'></monkey> <turtle name='luxy'></turtle></animals>")
	
	if rootNode then
		print("Trying to save non-file XML node")
		
		local saveXML = xmlSaveFile(rootNode)
		
		print("xmlSaveFile Result:", saveXML)
		
		print("Trying to unload non-file XML node")
		
		local unloadXML = xmlUnloadFile(rootNode)
		
		print("xmlUnloadFile Result:", unloadXML)
	end
end
addEventHandler("onClientResourceStart", resourceRoot, init)
```

Both will return false in the above example.

Lpsd committed:
Fix handling of non-file XML nodes

Ensures that XML file functions (xmlSaveFile, xmlUnloadFile) do not work with XML nodes created without a file (via xmLoadString) and returns correctly to reflect this.

qaisjp commented on Client/mods/deathmatch/logic/lua/CLuaMain.cpp:
```diff
@@ -370,29 +370,36 @@ CXMLNode* CLuaMain::ParseString(const char* strXmlContent)
-    if (!m_XMLFiles.empty())
-        m_XMLFiles.remove(pFile);
-    delete pFile;
+	if (!m_XMLFiles.empty()) {
```
should short circuit, like

```cpp
if (m_XMLFiles.empty()) {
    return false;
}

// ...
return true;
```

same for other functions

qaisjp commented on Client/mods/deathmatch/logic/lua/CLuaMain.cpp:
```diff
@@ -363,29 +363,43 @@ CXMLFile* CLuaMain::CreateXML(const char* szFilename, bool bUseIDs, bool bReadOn
+{
+	if (!m_XMLFiles.empty()) {
+		list<CXMLFile*>::iterator iter;
+		for (iter = m_XMLFiles.begin(); iter != m_XMLFiles.end(); iter++)
```
I think `for (CXMLFile* file : m_XMLFiles)` should work as a shorthand

qaisjp commented:
your code contains tabs, please indent using spaces

Lpsd committed:
Implement short-circuit suggestion

Lpsd commented:
Sorry! Removed the tabs.

The suggestion regarding the loop seems to crash the client.

Lpsd committed:
Untabify code

Lpsd committed:
Clean up SaveXML loop

Lpsd commented:
The iterator/index value of the loop within `CLuaMain::DestroyXML` is passed to `m_XMLFiles.erase`, is it worth the time simply for "readability"? (also, the loop was already like this, and I feel it falls outside the specification of my PR)

I have, however, cleaned up the loop in `CLuaMain::SaveXML` (one of them, at least)

Aside from this, everything seems to be working great.

Lpsd committed:
Fix node iteration & return errors on invalid XML string

Lpsd commented:
Fixed some issues brought to my attention by @patrikjuvonen 

XML strings are now properly parsed, to iterate through all children (instead of just the 1st)

Also added checks for broken XML strings, and error outputting

Lpsd committed:
Fix formatting in CXMLImpl::ParseString

patrikjuvonen committed:
Clean up a for loop and some misc small things

Co-Authored-By: LopSided <lpsd@users.noreply.github.com>

patrikjuvonen commented:
Waits for #821 to solve.

qaisjp commented:
@patrikjuvonen I don't think this should be assigned to 1.5.7

patrikjuvonen commented:
> @patrikjuvonen I don't think this should be assigned to 1.5.7

I've moved it to backlog now. We should probably check what else is left in 1.5.7 milestone and move them elsewhere if unlikely to get them completed in time.

botder commented:
The decision in pull request #821 (**Pin down return value on failure**) is now settled.

Lpsd committed:
Amend return value on failure

patrikjuvonen commented on Shared/mods/deathmatch/logic/luadefs/CLuaXMLDefs.cpp:
```diff
@@ -214,6 +216,37 @@ int CLuaXMLDefs::xmlLoadFile(lua_State* luaVM)
+    CScriptArgReader argStream(luaVM);
+    argStream.ReadString(strXmlContent);
+
+    if (!argStream.HasErrors())
```
You could simplify this slightly:
```cpp
if (argStream.HasErrors())
    return luaL_error(luaVM, argStream.GetFullErrorMessage());
```

Lpsd replied:
Done

Lpsd committed:
Implement requested change

patrikjuvonen commented:
Cool, thank you!
Great job on this PR. All ready to go after 1.5.7 is released! 👍
